types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// DynamicTypable extends attr.Type for dynamic types. Implement this interface to create a custom DynamicType type.
+type DynamicTypable interface {
+	attr.Type
+
+	// ValueFromDynamic should convert the DynamicValue to a DynamicValuable type.
+	ValueFromDynamic(context.Context, DynamicValue) (DynamicValuable, diag.Diagnostics)
+}
+
+var _ DynamicTypable = DynamicType{}
+
+// DynamicType is the base framework type for a dynamic. Static types are always
+// preferable over dynamic types in Terraform as practitioners will receive less
+// helpful configuration assistance from validation error diagnostics and editor
+// integrations.
+//
+// DynamicValue is the associated value type and, when known, contains a concrete
+// value of another framework type. (StringValue, ListValue, ObjectValue, MapValue, etc.)
+type DynamicType struct{}
+
+// MAINTAINER NOTE: Based on dynamic type alone, there is no alternative type information to return related to a path step.
+// When working with dynamics, we should always use DynamicValue to determine underlying type information.
+func (t DynamicType) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (interface{}, error) {
+	return nil, fmt.Errorf("cannot apply AttributePathStep %T to %s", step, t.String())
+}
```

Comment:
Nit: Do we want to move the maintainer code comment into the function body so it doesn't show up in the Go package documentation?